horace_core/sqw/@PixelData/PixelData.m

```diff
@@ -327,41 +440,110 @@
     end
 
     function detector_index = get.energy_idx(obj)
-       detector_index = obj.data(obj.FIELD_INDEX_MAP_('energy_idx'), :);
+        obj = obj.load_first_page_if_data_empty_();
+        detector_index = obj.data(obj.FIELD_INDEX_MAP_('energy_idx'), :);
     end
 
     function obj = set.energy_idx(obj, energies)
         obj.data(obj.FIELD_INDEX_MAP_('energy_idx'), :) = energies;
-     end
+    end
 
     function signal = get.signal(obj)
-       signal = obj.data(obj.FIELD_INDEX_MAP_('signal'), :);
+        obj = obj.load_first_page_if_data_empty_();
+        signal = obj.data(obj.FIELD_INDEX_MAP_('signal'), :);
     end
 
     function obj = set.signal(obj, signal)
         obj.data(obj.FIELD_INDEX_MAP_('signal'), :) = signal;
-     end
+    end
 
     function variance = get.variance(obj)
-       variance = obj.data(obj.FIELD_INDEX_MAP_('variance'), :);
+        obj = obj.load_first_page_if_data_empty_();
+        variance = obj.data(obj.FIELD_INDEX_MAP_('variance'), :);
     end
 
     function obj = set.variance(obj, variance)
         obj.data(obj.FIELD_INDEX_MAP_('variance'), :) = variance;
-     end
+    end
 
     function num_pix = get.num_pixels(obj)
-        num_pix = size(obj.data, 2);
+        if isempty(obj.f_accessor_)
+            num_pix = size(obj.data, 2);
+        else
+            num_pix = obj.f_accessor_.npixels;
+        end
+    end
+
+    function file_path = get.file_path(obj)
+        file_path = obj.file_path_;
+    end
+
+    function page_size = get.page_size(obj)
+        % The number of pixels that are held in the current page. If no pixels
+        % are held, this returns the number of pixels that will be loaded into
+        % the next page.
+        %
+        % If "num_pixels in file" < "number of pixels that can fit in memory size",
+        % return num_pixels - this avoids ever reading past end of file
+        if isempty(obj.data)
+            num_bytes_in_val = 8;  % pixel data stored in memory as a double
+            num_bytes_in_pixel = num_bytes_in_val*obj.PIXEL_BLOCK_COLS_;
+            page_size = floor(obj.page_memory_size_/num_bytes_in_pixel);
+            if page_size > obj.num_pixels
+                page_size = obj.num_pixels;
+            end
+        else
+            page_size = size(obj.data, 2);
+        end
+    end
+
+end
+
+methods (Access = private)
+
+    function obj = init_from_file_accessor_(obj, f_accessor)
+        % Initialise a PixelData object from a file accessor
+        obj.f_accessor_ = f_accessor;
+        obj.file_path_ = fullfile(obj.f_accessor_.filepath, ...
+                                  obj.f_accessor_.filename);
+        pg_size = obj.page_size;
+        if obj.num_pixels <= pg_size
+            obj.data = obj.f_accessor_.get_pix(1, pg_size);
+            obj.pix_position_ = obj.num_pixels;
+        end
     end
 
+    function obj = load_page_(obj, pix_idx_start)
+        if pix_idx_start >= obj.num_pixels
```

Comment:
Calling `.advance` won't work because as far as the API is concerned the first page is already there, so calling `.advance` will give you the second page. But I see your point, a user should never see a `PIXELDATA:load_page_`. Although that should never be thrown within `load_first_page_if_data_empty_` unless the SQW file has no pixels - is this possible?